### PR TITLE
Core: Add logLevel preset property to filter logging

### PR DESCRIPTION
**Pull request by ndelangen**
Issue: #8461

## What I did

ADD logLevel preset property (default is `info`)
ADD global variable injection
ADD filter for lower level log messages
ADD a global method for changing the logLevel at runtime.

### CODE REVIEW

File: addons/links/src/preview.test.js [hunk 1]
```diff
@@ -17,6 +17,8 @@ jest.mock('global', () => ({
       kind: 'kind',
     })),
   },
+  window: global,
+  __STORYBOOK_LOGGER: 'warn',
```
Comment (shilman) [inline addons/links/src/preview.test.js:21]: `__STORYBOOK_LOG_LEVEL__`?

Comment (shilman) [inline addons/links/src/preview.test.js:21]: what is this supposed to do? below `__STORYBOOK_LOGGER` is set to `console`, which is a different type?

File: lib/core/src/server/preview/iframe-webpack.config.js [hunk 2]
```diff
@@ -39,6 +39,7 @@ export default async ({
   presets,
 }) => {
   const dlls = await presets.apply('webpackDlls', []);
+  const logLevel = await presets.apply('logLevel', []);
```
Comment (shilman) [inline lib/core/src/server/preview/iframe-webpack.config.js:42]: Is this so the user can set `logLevel` in `main.js`? what happens when it gets defined in a preset? Is there one level globally, or does it get scoped somehow?

**Review by shilman — COMMENTED** [review]
LGTM but a few questions

File: lib/core/src/server/preview/iframe-webpack.config.js (hunk 2 above)
Comment (ndelangen) [reply lib/core/src/server/preview/iframe-webpack.config.js]: The value is a singular string, not an array.

There's 1 setting, whichever is last wins, `main.js` should always be last.

presets should not be setting this I think.

**Comment by ndelangen** [comment]
No idea why puppeteer is all of a sudden not able to find the globals anymore, and use them to extract.

When I run it locally it works.

File: lib/client-logger/src/index.ts [hunk 3]
```diff
@@ -1,16 +1,25 @@
...
-
-/* tslint:disable: no-console */
+import { window, LOGLEVEL } from 'global';
+import console from 'loglevel';
 
 export const logger = {
+  trace: (message: any, ...rest: any[]): void => console.trace(message, ...rest),
   debug: (message: any, ...rest: any[]): void => console.debug(message, ...rest),
-  log: (message: any, ...rest: any[]): void => console.log(message, ...rest),
+  log: (message: any, ...rest: any[]): void => window.console.log(message, ...rest),
```
Comment (tmeasday) [inline lib/client-logger/src/index.ts:7]: This is pretty surprising, what's up with this?

File: lib/client-logger/src/index.ts [hunk 4]
```diff
@@ -1,16 +1,25 @@
...
   error: (message: any, ...rest: any[]): void => console.error(message, ...rest),
 };
 
-export const pretty = (type: keyof typeof logger) => (...args: string[]) => {
+if (LOGLEVEL) {
+  console.setLevel(LOGLEVEL);
+}
+
+// eslint-disable-next-line no-underscore-dangle
+window.__STORYBOOK_LOGGER = console;
```
Comment (tmeasday) [inline lib/client-logger/src/index.ts:18]: Should we just check if `window` is defined before setting this?

**Comment by tmeasday** [comment]
@ndelangen I think the most likely situation is that the storybook is throwing an error as it loads in puppeteer and we aren't seeing the error. Try adding:

```js
  this.page.on('console', (msg) => {
    const type = msg.type();
    logger.info({ browserConsole: true, type, ...msg.location() }, msg.text());
  });
```

Either that or get a build storybook off one of the CI machines (I have one here)

File: lib/client-logger/src/index.ts (hunk 4 above)
Comment (ndelangen) [reply lib/client-logger/src/index.ts:18]: ✅

File: lib/client-logger/src/index.ts (hunk 3 above)
Comment (ndelangen) [reply lib/client-logger/src/index.ts:7]: `log` has no "level" to it. it's not part of `console` lib.

I could decide to make `log` equal to `info`?

**Review by shilman — APPROVED** [review]
LGTM does this need documentation somewhere?

**Comment by ndelangen** [comment]
I added documentation and the ability to set a value for this from the cli

**Pull request merged**